src/pgtable.c

```diff
@@ -0,0 +1,588 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
+ * Copyright (C) 2022 Amazon.com, Inc. or its affiliates.
+ *
+ * See file LICENSE for terms.
+ */
```

Comment:
This seems overly complex for what we need, with callbacks where simple function calls should be use instead. We already have code in tree for managing a cache.  Why is that code not used?

Reply:
I will re-work the PR to use rbmap which is used when managing mr cache. That will get rid of the current pgtable stuff that is copied from UCX, and also gets rid of the license concern we have.

Reply:
Is the rbmap as efficient as the pgtable or is their room for improvement?
The first commit hints at it being fairly optimized for their use case.
https://github.com/openucx/ucx/commit/9dbd386eb3f08bb0512eb39bb52450bfcf75f0fa

Reply:
I can probably run a comparison if we have an implementation with rbmap. 

When I adopted the pgtable caching my tests showed a good performance boost for SHM testing, because we no longer had to open/close the shared handled on every write/read, which were pretty expensive. So the first time, it would get a cache miss, then on every other iteration of the OSU test it would be a cache hit. However, most of the performance was obtained when I moved to asynchronous memory copies, which I implemented for ROCM. I did an implementation for both HSA and HIP. I'm assuming HIP style asynchronous memcpy can work for CUDA (just an educated guess, never looked at cuda)

The difference in performance between the pgtable and rbmap implementation would be in the handle lookup.  Note, I also used the same mechanism for XPMEM support

Reply:
@amirshehataornl I observed 30x time perf boost when switching from no-cache to cache with the current pgtable implementation, for CUDA IPC when running shm test. I can redo such perf test after I implement the cache with rbmap.

Reply:
@shijin-aws thanks for the info.. I'd like to do the same for rocm

Reply:
The ofi_mr_cache currently uses an rbtree, but that's hidden from the user.  There's no reason why it couldn't change to some other lookup method.  My question was is there a reason that cannot be used, either as-is or by modifying it?  There's already support for hmem as part of the ofi_mr_entry.

I completely believe that a cache is faster.

Reply:
@shefty Yeah you are right, we should be able to reuse the same lookup method for the ipc cache. One question raised earlier is whether this lookup method (rbtree) is more performant than the pgtable in UCX. We can run some comparison for it.

Reply:
The rbtree implementation was just easy.  We can switch to a hash table or other lookup method.  It makes sense to me to start with what we have and update the ofi_mr_cache separately to improve the performance.  That way, all cached entries benefit from a faster lookup, rather than only IPC related GPU buffers.

Reply:
I am a bit uncertain whether we should use `ofi_mr_cache` for IPC cache. They are different things right? You don't need a memory registration to use IPC? and I don't see whether a memory monitor is really needed for IPC cache. The IPC handle being opened must be valid because it's just got by the peer process.

Reply:
We are caching memory regions (identified by an iovec -- address + length), and storing with the region some metadata.  The current metadata includes the iovec and possible device interface and id, plus some provider specific data.  What we use the cache for doesn't really matter.  It's an address + length lookup to return some data structure.  If the memory monitor always returns that everything in the cache is always valid, then that's easy.  It would definitely be surprising for items that are cached, but if it works, that's fine.

Reply:
Makes sense, thanks for the clarification!

Reply:
@shefty If we're going to use the mr cache as a generic iov cache, I wonder if this will necessitate the bringing back of merge regions as an option of the cache

Reply:
Maybe, but I doubt it will be necessary.  Merging resulted in worse performance when we actually needed to do something as part of the merge (for example, register a new memory region to replace the ones being merged).

I thought in the case of HMEM, we retrieved the base address of the allocated region using the target address as input.  So, the cached region should be relative to the base address, not the address passed in by the app.

Reply:
In the design I mocked up to see how it would work, I set the stored base addr and len to be the base address of an allocation and 1 (as that would be how CUDA works), in which case merge regions wouldn't be needed. However, I didn't know if all ipc handles maintain the same behavior of base ptr address. The UCX implementation for example does base and len, and invalidates regions based on overlap iirc. If we can just use [base_addr, 1] as the key, then it shouldn't be too difficult to do.

Reply:
I'm guessing there's no way to get the actual size of the allocated region, just the base address?

Reply:
for cuda, you can do cuPointerGetAttribute, to get the base address, and the length?

Reply:
The actual allocation length is also available with oneAPI L0.